design-document.md

```diff
@@ -232,6 +232,352 @@ Note: Two non-colluding aggregators --- one leader and one standard aggregator
+following payload:
+
+```
+struct {
+  PATask task;                // PAParameter.task
+  uint8 helper_key_config_id; // HPKEKeyConfig.id
+  Url helper_url;             // URL of helper endpoint
+  PAHelperShare helper_share; // The helper's encrypted share
+  PALeaderShare leader_share; // The leader's share
+} PAUplaod;
+
+struct {
+  opaque enc<0..2^16-1>;     // Encapsulated HPKE context
+  opaque payload<0..2^24-1>; // The encrypted share
+} PAHelperShare;
+
+opaque PALeaderShare<0..2^24-1>;
+```
+
+### Input validation
+
+The leader handles HTTP requests to `[leader]/upload` as follows. It first
+checks the request for a well-formed `PAUpload` message. If not found, then the
+leader aborts and alerts the client with "malformed upload". Otherwise, if a
+well-formed `PAUpload` message is found, the leader responds to the request with
+"200 OK".
+
+#### Verify Start
+
+The leader begins the input-validation protocol by looking up the PA parameters
+`PAParam` for which `PAUpload.task == PAParam.task` and `PAUpload.helper_url` is
+in `PAParam.helper_urls`. If none are found, then the leader aborts without
+alerting the client.
+
+**Request.**
+Let `[helper]` denote the URL corresponding to `PAUpload.helper_url`. The leader
+sends an HTTP request to `[helper]/verify_start` with the following payload:
+
+```
+struct {
+  PATask task;                           // PAUpload.task
+  uint8 key_config_id;                   // PAUpload.helper_key_config_id
+  PAHelperShare shares<0..2^24-1>;       // Any number of helper shares
+  select (PAVerifyStartReq.task.proto) { // Payload for each helper share
+    case prio: PrioVerifyStartReq payloads<0..2^16-1>;
+    case hits: HitsVerifyStartReq payloads<0..2^16-1>;
+  }
+} PAVerifyStartReq;
+```
+
+This message contains any number of encrypted helper shares: it might contain
+the shares uploaded by a single client, or it might contain the shares uploaded
+by multiple clients; we only require that all of the shares correspond to the
+same PA task and HPKE key configuration. This allows the aggregator to
+batch-process a number of inputs at once.
+
+Each helper share has a corresponding message payload contained in
+`PAVerifyStartReq.shares`. This includes any protocol-specific information the
+helper needs for input validation, e.g., the joint randomness used by the leader
+and helper for the protocol run.
+
+Note that a well-formed message contains as many payloads as shares, i.e., the
+sequence `PAVerifyStartReq.payloads` has the same number of elements as
+`PAVerifyStartReq.shares`. Moreover, the i-th payload should correspond to the
+i-th share for each i.
+
+**Response.**
+The helper handles HTTP requests to `[helper]/verify_start` as follows. It first
+checks for a well-formed `PAVerifyStartReq`. If not found, then the helper
+aborts and alerts the leader with "malformed verify start request".
+
+Next, it looks for PA parameters `PAParam` for which `PAVerifyStartReq.task ==
```

Comment:
Looks where? I think the idea is that each helper maintains a list of `PAParam`s it is aware of and participating in?

Reply:
Yup, that's the idea.

Reply:
I think it'd help to have language explaining that all participants are configured with a list of known `PAParam`s before the start of the discovery or ingestion protocols.

Reply:
Done. (See the changes to the beginning of {{proto-ingestion}}.